ext/CEGUI/IrrlichtBaWTexture.cpp

```diff
@@ -0,0 +1,167 @@
+    auto bundle = assetManager->getAsset(path, params);
+    auto cpuImage = asset::IAsset::castDown<asset::ICPUImage>(bundle.getContents().first[0]);
+    assert(cpuImage.get(), "something went wrong while casting, empty pointer detected!");
+
+    bufferRowLength = cpuImage->getRegions().begin()->bufferRowLength;
+
+    asset::ICPUImageView::SCreationParams imageViewInfo;
+    imageViewInfo.image = std::move(cpuImage);
+    imageViewInfo.format = imageViewInfo.image->getCreationParameters().format;
+    imageViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+    imageViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+    imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+    imageViewInfo.subresourceRange.baseMipLevel = 0u;
+    imageViewInfo.subresourceRange.layerCount = imageViewInfo.image->getCreationParameters().arrayLayers;
+    imageViewInfo.subresourceRange.levelCount = imageViewInfo.image->getCreationParameters().mipLevels;
+
+    auto cpuImageView = asset::ICPUImageView::create(std::move(imageViewInfo));
+
+    // cachingName = ""     TODO Okay, so we have to get likely GL id or something similar. Should I cast it to OpenGL image or maybe we should get CPU cache string?
+
+    gpuTexture = device->getVideoDriver()->getGPUObjectsFromAssets(&cpuImageView.get(), &cpuImageView.get() + 1u)->front();
+    
+}
+
+void IrrlichtBaWTexture::loadFromMemory(const void* buffer, const CEGUI::Sizef& buffer_size, PixelFormat pixel_format)
+{
+    asset::E_FORMAT newFormat = [&]()
+    {
+        switch (pixel_format)
+        {
+            case PixelFormat::PF_RGB: return asset::EF_R8G8B8_SRGB;
+            case PixelFormat::PF_RGBA: return asset::EF_R8G8B8A8_SRGB;
+            case PixelFormat::PF_RGBA_4444: return asset::EF_R4G4B4A4_UNORM_PACK16;
+            case PixelFormat::PF_RGB_565: return asset::EF_R5G6B5_UNORM_PACK16;
+            case PixelFormat::PF_PVRTC2: return asset::EF_PVRTC2_2BPP_SRGB_BLOCK_IMG;
+            case PixelFormat::PF_PVRTC4: return asset::EF_PVRTC2_4BPP_SRGB_BLOCK_IMG;
+            case PixelFormat::PF_RGB_DXT1: return asset::EF_BC1_RGB_SRGB_BLOCK;
+            case PixelFormat::PF_RGBA_DXT1: return asset::EF_BC1_RGBA_SRGB_BLOCK;
+            case PixelFormat::PF_RGBA_DXT3: return asset::EF_BC2_SRGB_BLOCK;
+            case PixelFormat::PF_RGBA_DXT5: return asset::EF_BC3_SRGB_BLOCK;
+                
+            default:
+                return asset::EF_UNKNOWN;
+        }
+    }();
+
+    asset::IImage::SCreationParams imageParams;
+    imageParams.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0);
+    imageParams.type = asset::IImage::ET_2D;
+    imageParams.format = newFormat;
+    imageParams.extent = { buffer_size.d_width, buffer_size.d_height, 1 };
+    imageParams.mipLevels = 1;
+    imageParams.arrayLayers = 1;
+
+    auto gpuImage = device->getVideoDriver()->createGPUImage(std::move(imageParams));
+    // TODO copy buffer to image
```

Comment:
call `blitFromMemory(buffer,{0.f,0.f,buffer_size.d_width, buffer_size.d_height})` and do the implementation there

Reply:
changed it a little